### PR TITLE
For #17798: Sync Menu Item

### PR DESCRIPTION
**Pull request by kglazko**
For #17798 

### Pull Request checklist

- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Screenshots**: This PR includes screenshots or GIFs of the changes made or an explanation of why it does not
- [ ] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features. In addition, it includes a screenshot of a successful [accessibility scan](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor&hl=en_US) to ensure no new defects are added to the product.

### To download an APK when reviewing a PR:
1. click on Show All Checks,
2. click Details next to "Taskcluster (pull_request)" after it appears and then finishes with a green checkmark,
3. click on the "Fenix - assemble" task, then click "Run Artifacts".
4. the APK links should be on the left side of the screen, named for each CPU architecture

### CODE REVIEW

**Comment by kglazko** [comment]
Putting this up for feedback @eliserichards @gabrielluong just to see what you think of this approach. I see a bug with this on the emulator where for some reason on the emulator when you install it is never recognized as signed in until you navigate somewhere in the app and come back. Attaching a photo to show how it looks.
Screen Shot 2021-03-23 at 1 21 38 PM

**Comment by eliserichards** [comment]
What are the steps that get to this point? Open the app (fresh build), open menu to settings, sign in...

**Comment by kglazko** [comment]
Yeah, I would have to go to Settings, sign into Sync, and then see whether or not it shows up on the menu. I signed into Sync when it still wasn't working and it kept me signed in so eventually I would see it on a fresh install of the app- but only after navigating somewhere else first, I think the menu loads before Sync does maybe...

**Review by eliserichards — COMMENTED** [review]
This is a good start! A few pointers: 

* Make sure we're adding this in both the `HomeMenu` (for new tab) and the `DefaultToolbarMenu` (main menu)
* Using the AccountManager is the right way to go. Take a look at the [SyncPreferenceView](https://github.com/mozilla-mobile/fenix/blob/master/app/src/main/java/org/mozilla/fenix/settings/SyncPreferenceView.kt) and how it's registering for account changes from the manager. We want to be listening for any changes and dynamically updating the list when we receive a change
* The [AccountUiView](https://github.com/mozilla-mobile/fenix/blob/master/app/src/main/java/org/mozilla/fenix/settings/account/AccountUiView.kt) might also be helpful
* When we [receive an update](https://github.com/mozilla-mobile/fenix/blob/246c192de377fa180029604846bb3497c4770dd9/app/src/main/java/org/mozilla/fenix/components/toolbar/DefaultToolbarMenu.kt#L532) from the `AccountManager`, we want to update the item in our menu. Can we dynamically update _just the item_, or do we need to refresh the whole list? It looks like we're already doing a rebuild of the list here: https://github.com/mozilla-mobile/fenix/blob/246c192de377fa180029604846bb3497c4770dd9/app/src/main/java/org/mozilla/fenix/home/HomeMenu.kt#L243

* > I think the menu loads before Sync does
   * If we are loading too quickly, we want to make sure we wait until all processes are complete before rebuilding the list

**Comment by kglazko** [comment]
Yeah, upon investigating it looks like we are already rebuilding the list on a log out or reuthenticate event. So that's good. I also realized that I couldn't find finished UI tests around sync sign-in, there are some that are skeleton-ed, but we may need to write them for this PR.

**Comment by kglazko** [comment]
>       * If we are loading too quickly, we want to make sure we wait until all processes are complete before rebuilding the list

Do you  have any ideas on how we can do this, @eliserichards ? We probably wouldn't want to hold up the rest of the menu UI from loading if it's waiting on a process.

**Comment by eliserichards** [comment]
For the UI tests, fix up what you can and then @ignore the ones that need more time and Isabel's team will look at them. Make sure you tag the UI test issue in the @ignore string so they know where to look.

WRT loading: you're correct.. the menu should load and shouldn't be waiting on the sync. So it should default to the signed out UI state and then reload everything once we get the auth response

File: app/src/androidTest/java/org/mozilla/fenix/ui/robots/SyncedTabsRobot.kt [hunk 1]
```diff
@@ -39,6 +39,8 @@ private fun goBackButton() =
     onView(allOf(withContentDescription("Navigate up")))
 
 private fun assertSyncedTabsMenuHeader() {
```
Comment (eliserichards) [inline app/src/androidTest/java/org/mozilla/fenix/ui/robots/SyncedTabsRobot.kt:41]: Nit: rename the test something like this so you don't need the inline comment: 
```suggestion
 private fun `GIVEN xxxx WHEN xxxx THEN xxxx`() {
```

File: app/src/main/java/org/mozilla/fenix/components/toolbar/DefaultToolbarMenu.kt [hunk 2]
```diff
@@ -236,10 +236,18 @@ class DefaultToolbarMenu(
             onItemTapped.invoke(ToolbarMenu.Item.AddToHomeScreen)
         }
 
+        val accountManager = context.components.backgroundServices.accountManager
+        val notSignedIn = accountManager.authenticatedAccount()
```
Comment (eliserichards) [inline app/src/main/java/org/mozilla/fenix/components/toolbar/DefaultToolbarMenu.kt:240]: Is this an async call? Do we need to ensure that this has finished before evaluating `notSignedIn`?

Comment (eliserichards) [reply app/src/main/java/org/mozilla/fenix/components/toolbar/DefaultToolbarMenu.kt:240]: Also, should this be `signedIn` instead?

File: app/src/main/java/org/mozilla/fenix/components/toolbar/DefaultToolbarMenu.kt [hunk 3]
```diff
@@ -236,10 +236,18 @@ class DefaultToolbarMenu(
...
+        val syncStringToDisplay = if (notSignedIn != null && accountManager.accountProfile()?.email != null) {
```
Comment (eliserichards) [inline app/src/main/java/org/mozilla/fenix/components/toolbar/DefaultToolbarMenu.kt:241]: Nit:
```suggestion
        val syncItemTitle
```

File: app/src/main/res/values/strings.xml [hunk 4]
```diff
@@ -953,6 +953,8 @@
     <string name="share_link_all_apps_subheader">All actions</string>
     <!-- Sub-header in the dialog to share a link to an app from the most-recent sorted list -->
     <string name="share_link_recent_apps_subheader">Recently used</string>
+    <!-- An string shown when an account is signed in where %1$s is a placeholder for the email-->
+    <string name="sync_signed_as">Signed in as %1$s</string>
     <!-- An option from the share dialog to sign into sync -->
     <string name="sync_sign_in">Sign in to Sync</string>
```
Comment (eliserichards) [inline app/src/main/res/values/strings.xml:957]: Just confirming this is capital S "Sync" not lowercase 😁

**Review by eliserichards — CHANGES_REQUESTED** [review]
Looking good! A few questions and nits. Make sure we're only changing the synced tabs item in `newCoreMenuItems`, not the old menu items.

File: app/src/main/java/org/mozilla/fenix/components/toolbar/DefaultToolbarMenu.kt (hunk 2 above)
Comment (gabrielluong) [reply app/src/main/java/org/mozilla/fenix/components/toolbar/DefaultToolbarMenu.kt:240]: I think it probably makes more sense to call it `account` to not confuse this with a boolean value.

File: app/src/main/res/values/strings.xml (hunk 4 above)
Comment (kglazko) [reply app/src/main/res/values/strings.xml:957]: Good catch, the design spec shows it as lowercase! Changing it now :)

File: app/src/androidTest/java/org/mozilla/fenix/ui/robots/SyncedTabsRobot.kt (hunk 1 above)
Comment (kglazko) [reply app/src/androidTest/java/org/mozilla/fenix/ui/robots/SyncedTabsRobot.kt:41]: For UI tests, the convention doesn't seem to be 'GIVEN WHEN THEN' but rather the assert style based on a few other robots I looked at. Tagging @AaronMT or @rpappalax to give input

**Comment by kglazko** [comment]
Thank you for your feedback, @eliserichards and @gabrielluong. I incorporated your comments, let me know if there's anything else I can do.

File: app/src/androidTest/java/org/mozilla/fenix/ui/robots/SyncedTabsRobot.kt (hunk 1 above)
Comment (eliserichards) [reply app/src/androidTest/java/org/mozilla/fenix/ui/robots/SyncedTabsRobot.kt:41]: that's totally fine, I didn't realize 😄

**Comment by eliserichards** [comment]
One thing to confirm: when we click on the item, where do we want it to go? right now it's directing to our synced tabs. We can confirm this with Victoria

**Comment by kglazko** [comment]
Filing this as a follow-up after discussing with @eliserichards https://github.com/mozilla-mobile/fenix/issues/18806

**Review by eliserichards — APPROVED** [review]
Awesome job! 🎉

File: app/src/main/java/org/mozilla/fenix/components/toolbar/DefaultToolbarMenu.kt [hunk 5]
```diff
@@ -402,13 +402,21 @@ open class DefaultToolbarMenu(
...
+            context.getString(R.string.sync_menu_sign_in)
+        }
+
+        val syncedTabs = BrowserMenuImageText(
+            syncItemTitle,
+            R.drawable.ic_synced_tabs,
+            primaryTextColor()
+        ) {
+            onItemTapped.invoke(ToolbarMenu.Item.SyncedTabs)
+        }
```
Comment (gabrielluong) [inline app/src/main/java/org/mozilla/fenix/components/toolbar/DefaultToolbarMenu.kt:419]: I think we got an extra indent here.

File: app/src/main/java/org/mozilla/fenix/home/HomeMenu.kt [hunk 6]
```diff
@@ -215,13 +223,11 @@ class HomeMenu(
...
         // We don't want to cause its initialization just for this check.
-        val accountAuthItem =
-            if (context.components.backgroundServices.accountManagerAvailableQueue.isReady() &&
-                context.components.backgroundServices.accountManager.accountNeedsReauth()) {
-                    reconnectToSyncItem
-            } else {
-                null
-            }
+        val accountAuthItem = if (context.components.backgroundServices.accountManagerAvailableQueue.isReady()) {
+            if (context.components.backgroundServices.accountManager.accountNeedsReauth()) reconnectToSyncItem else null
```
Comment (gabrielluong) [inline app/src/main/java/org/mozilla/fenix/home/HomeMenu.kt:227]: Just for awareness I think this extra if statement would be the equivalent as the above using &&. This probably didn't need to be change.

**Pull request merged**